src/vs/code/node/cli.ts

```diff
@@ -319,23 +319,89 @@ export async function main(argv: string[]): Promise<any> {
 			options['stdio'] = 'ignore';
 		}
 
-		const child = spawn(process.execPath, argv.slice(2), options);
+		let child: ChildProcess;
+		if (platform() !== 'darwin') {
+			child = spawn(process.execPath, argv.slice(2), options);
 
-		if (args.wait && waitMarkerFilePath) {
-			return new Promise<void>(resolve => {
+			if (args.wait && waitMarkerFilePath) {
+				return new Promise<void>(resolve => {
 
-				// Complete when process exits
-				child.once('exit', () => resolve(undefined));
+					// Complete when process exits
+					child.once('exit', () => resolve(undefined));
 
-				// Complete when wait marker file is deleted
-				whenDeleted(waitMarkerFilePath!).then(resolve, resolve);
-			}).then(() => {
+					// Complete when wait marker file is deleted
+					whenDeleted(waitMarkerFilePath!).then(resolve, resolve);
+				}).then(() => {
 
-				// Make sure to delete the tmp stdin file if we have any
-				if (stdinFilePath) {
-					unlinkSync(stdinFilePath);
+					// Make sure to delete the tmp stdin file if we have any
+					if (stdinFilePath) {
+						unlinkSync(stdinFilePath);
+					}
+				});
+			}
+		} else {
+			const requiresWait = args['status'] || args['verbose'] || hasReadStdinArg;
+			const envVars: string[] = [];
+			for (const envKey in env) {
+				if (envKey === '_') {
+					continue;
 				}
-			});
+				const value = env[envKey];
+				envVars.push('--env');
+				envVars.push(envKey + '=' + value ?? '');
+			}
+			const openArgs: string[] = ['-n'];
+
+			let tmpfile = '';
+			let tmpfileError = '';
+			if (requiresWait) {
+				const time = new Date().getTime();
+				tmpfile = `${tmpdir()}/vscode-wait-transfer-${time}.log`;
+				tmpfileError = `${tmpdir()}/vscode-wait-transfer-error-${time}.log`;
+				writeFileSync(tmpfile, '');
+				openArgs.push('-W');
+				openArgs.push('--stdout', tmpfile);
+				openArgs.push('--stderr', tmpfileError);
+				argv.push('--wait');
+			}
+			const argsArr: string[] = [];
+			const execPathToUse = args['exec-path'] ?? process.execPath;
+			argsArr.push('-a', execPathToUse);
+			argsArr.push(...envVars, ...openArgs, '--args', ...argv.slice(2));
+			child = spawn('open', argsArr, options);
```

Comment:
Do we need to pass the `env` in options to both `spawn` and into the `open` command? Or is one sufficient?

Reply:
The one passed to the `open` command is sufficient, because open expects its env variables in a specific format then `env` passed to spawn will be noop, we can remove it.

Reply:
I just verified that the one provided as part of the `options` parameter is necessary and sufficient. I'm not sure why with `code-cli.sh` that works, but with the `--exec-path` approach, I remember needing to pass in the env as args .

Reply:
I am confused, from `man open`

```
--env VAR
         Adds VAR to the environment of the launched application.  VAR should be formatted NAME=VALUE or NAME.
```

I expect whatever `env` that is passed to `child_process.spawn` options is not propagated by open unless `--env` arg is used as part of the open command. Since `open` executes the application via launchctl it will not inherit any environment from whatever parent process that called the `open` command.

Reply:
I just took another look, and there is another line earlier that says `Opened applications inherit environment variables just as if you had launched the application directly through its full path.`

It seems that both `VSCODE_DEV=1 open -a /Users/raymondzhao/work/vscode/.build/electron/Code\ -\ OSS.app/ --args /Users/raymondzhao/work/vscode` and `open -a /Users/raymondzhao/work/vscode/.build/electron/Code\ -\ OSS.app/ --env VSCODE_DEV=1 --args /Users/raymondzhao/work/vscode` work on my machine, though by passing in the JSON object, I avoid having to manually convert over the environment variables or deal with the `_` environment variable.

Reply:
Here's something interesting that I just encountered: it seems when `--env` is used instead, `open` does a check to see whether or not the vars can be applied:

```
raymondzhao@Raymonds-MacBook-Pro vscode % VSCODE_DEV= open -a /Users/raymondzhao/work/vscode/.build/electron/Code\ -\ OSS.app/ --args /Users/raymondzhao/work/vscode
raymondzhao@Raymonds-MacBook-Pro vscode % open -a /Users/raymondzhao/work/vscode/.build/electron/Code\ -\ OSS.app/ --env VSCODE_DEV=0 --args /Users/raymondzhao/work/vscode
Application /Users/raymondzhao/work/vscode/.build/electron/Code - OSS.app was already running and so the additional environment variables could not be set.
```

However, we use `-n` to always launch another instance and let the single instance lock handle the rest, so that skips over this issue.

Reply:
Yeah looks like they do inherit, but I cannot find the open source code so this is just going purely based on testing and looking at the documentation of the LaunchService api https://developer.apple.com/documentation/appkit/nsworkspace/openconfiguration that does allow passing in environments.